lib/spack/spack/directives.py

```diff
@@ -119,56 +170,39 @@ class Foo(Package):
     packages to the core.
 
     """
+    global _directive_names, __all__
 
-    def __init__(self, dicts=None):
-        if isinstance(dicts, basestring):
-            dicts = (dicts, )
-        elif type(dicts) not in (list, tuple):
-            raise TypeError(
-                "dicts arg must be list, tuple, or string. Found %s" %
-                type(dicts))
-
-        self.dicts = dicts
-
-    def ensure_dicts(self, pkg):
-        """Ensure that a package has the dicts required by this directive."""
-        for d in self.dicts:
-            if not hasattr(pkg, d):
-                setattr(pkg, d, {})
-
-            attr = getattr(pkg, d)
-            if not isinstance(attr, dict):
-                raise spack.error.SpackError(
-                    "Package %s has non-dict %s attribute!" % (pkg, d))
+    if isinstance(dicts, basestring):
+        dicts = (dicts, )
+    if not isinstance(dicts, collections.Sequence):
+        message = "dicts arg must be list, tuple, or string. Found {0}"
+        raise TypeError(message.format(type(dicts)))
+    # Add the dictionary names if not already there
+    _directive_names |= set(dicts)
 
-    def __call__(self, directive_function):
-        directives[directive_function.__name__] = self
-
-        @functools.wraps(directive_function)
-        def wrapped(*args, **kwargs):
-            pkg = DictWrapper(caller_locals())
-            self.ensure_dicts(pkg)
-
-            pkg.name = get_calling_module_name()
-            return directive_function(pkg, *args, **kwargs)
-
-        return wrapped
+    # This decorator just returns the directive functions
+    def _decorator(decorated_function):
+        __all__.append(decorated_function.__name__)
+        return decorated_function
+    return _decorator
 
 
 @directive('versions')
-def version(pkg, ver, checksum=None, **kwargs):
+def version(ver, checksum=None, **kwargs):
     """Adds a version and metadata describing how to fetch it.
-       Metadata is just stored as a dict in the package's versions
-       dictionary.  Package must turn it into a valid fetch strategy
-       later.
+    Metadata is just stored as a dict in the package's versions
+    dictionary.  Package must turn it into a valid fetch strategy
+    later.
     """
-    # TODO: checksum vs md5 distinction is confusing -- fix this.
-    # special case checksum for backward compatibility
-    if checksum:
-        kwargs['md5'] = checksum
+    def _execute(pkg):
+        # TODO: checksum vs md5 distinction is confusing -- fix this.
+        # special case checksum for backward compatibility
+        if checksum:
+            kwargs['md5'] = checksum
 
-    # Store kwargs for the package to later with a fetch_strategy.
-    pkg.versions[Version(ver)] = kwargs
+        # Store kwargs for the package to later with a fetch_strategy.
+        pkg.versions[Version(ver)] = kwargs
+    _directives_to_be_executed.append(_execute)
```

Comment:
Could the `directive` decorator either:

1. Pass this list in as a parameter (kind of like how `pkg` was passed in before)? or
2. Make directives return either something callable on a `Package`  or a list of them?

Either of these would make it so that not all directives have to refer to a global, and they'd consolidate any use of globals to the `directive` definition.   It would also make it clearer to potential contributors how they might write their own directive (as the interface would be clear from the structure).